### PR TITLE
api: add hardcoded versioning support

### PR DESCRIPTION
**Pull request by GRISHNOV**
Add versioning support, now the module api has _VERSION hardcoded variable

### CODE REVIEW

File: queue-scm-1.rockspec [hunk 1]
```diff
@@ -27,7 +27,8 @@ build = {
         ['queue.abstract.driver.limfifottl'] = 'queue/abstract/driver/limfifottl.lua',
         ['queue.compat']                     = 'queue/compat.lua',
         ['queue.util']                       = 'queue/util.lua',
-        ['queue']                            = 'queue/init.lua'
+        ['queue']                            = 'queue/init.lua',
+        ['queue-version']                    = 'queue-version.lua'
```
Comment (DifferentialOrange) [inline queue-scm-1.rockspec:31]: Let's use `queue/version.lua`.

For modules like `checks`, I also suggest to create a new folder `checks` and a file named `checks/version.lua`.

File: queue/init.lua [hunk 2]
```diff
@@ -45,6 +45,12 @@ queue = setmetatable({
     end
 })
 
+local ok, VERSION = pcall(require, 'queue-version')
```
Comment (DifferentialOrange) [inline queue/init.lua:48]: Well, for now there isn't any valid case when you can't require version file.

File: queue-scm-1.rockspec [hunk 3]
```diff
@@ -27,7 +27,8 @@ build = {
         ['queue.abstract.driver.limfifottl'] = 'queue/abstract/driver/limfifottl.lua',
```
Comment (DifferentialOrange) [inline queue-scm-1.rockspec:27]: Let's also start with CI check draft here, this things must be pushed together.

File: queue-version.lua [hunk 4]
```diff
@@ -0,0 +1,2 @@
+-- Сontains the module version, requires manual update.
```
Comment (DifferentialOrange) [inline queue-version.lua:1]: So what's our proposed plan here? Bump the version once per tag? Or something a bit more complicated like https://github.com/tarantool/metrics/pull/426/commits/b63093938a2988611bd568220fae6281f106c493 ?

File: queue-scm-1.rockspec (hunk 1 above)
Comment (GRISHNOV) [reply queue-scm-1.rockspec:31]: Ok, but we already have `vshard/version.lua` in vshard [repo](https://github.com/tarantool/vshard/tree/master/vshard). Then for `vshard` the name or location of version file will be different 

> Let's use queue/version.lua

ok, I thought this file would be more noticeable in the root of the repository and it would be harder to lose it

File: queue/init.lua (hunk 2 above)
Comment (GRISHNOV) [reply queue/init.lua:48]: It's more just in case. For example, some repositories use `make/cmake`, where you can forget to transfer this file to the installation location. Specifically, here it was possible to forget to add this file to `build.module` in the rockspec.
Well, or is it in case the user wants to transfer files to his project and forgets file with version
But I think it can be simplified without `pcall`

File: queue-version.lua (hunk 4 above)
Comment (GRISHNOV) [reply queue-version.lua]: Let's use https://github.com/tarantool/metrics/commit/b63093938a2988611bd568220fae6281f106c493  approach

File: queue-scm-1.rockspec (hunk 3 above)
[the remark this answers is not in the document]
Comment (GRISHNOV) [reply queue-scm-1.rockspec:27]: Done, now it is `api`. Commit description updated

File: queue-version.lua (hunk 4 above)
Comment (DifferentialOrange) [reply queue-version.lua]: We've discussed the question with @LeonidVas and decided to stick to the simplest approach -- just bump the version for a tag commit, do not mark scm version in any way.

Comment (GRISHNOV) [reply queue-version.lua]: Done

File: queue-scm-1.rockspec (hunk 3 above)
Comment (GRISHNOV) [reply queue-scm-1.rockspec:27]: Implemented job `hardcoded_ver_check` in `fast_testing.yml`

File: queue/init.lua [hunk 5]
```diff
@@ -45,6 +45,8 @@ queue = setmetatable({
     end
 })
 
+queue['_VERSION'] = require('queue.version')
```
Comment (DifferentialOrange) [inline queue/init.lua:48]: Let's put it in the `queue` table after `cfg`.

File: .github/workflows/fast_testing.yml [hunk 6]
```diff
@@ -10,6 +10,35 @@ on:
...
+        uses: actions/checkout@v2
+
+      - name: Fetch tags
+        # Found that Github checkout Actions pulls all the tags, but
+        # right it deannotates the testing tag, check:
+        #   https://github.com/actions/checkout/issues/290
+        # But we use 'git describe ..' calls w/o '--tags' flag and it
+        # prevents us from getting the needed tag for packages version
+        # setup. To avoid of it, let's fetch it manually, to be sure
+        # that all tags will exists always.
```
[the remark this answers is not in the document]
Comment (GRISHNOV) [reply .github/workflows/fast_testing.yml]: Added

File: .github/workflows/fast_testing.yml [hunk 7]
```diff
@@ -10,6 +10,35 @@ on:
...
+        run: git fetch --tags -f
+
+      - name: Fetch hardcoded and last repository tag versions
+        run: |
+          echo HARDCODED_VERSION=$(cat queue/version.lua | grep -oP "'\K[^']+" ) >> $GITHUB_ENV
```
[the remark this answers is not in the document]
Comment (GRISHNOV) [reply .github/workflows/fast_testing.yml]: Done

File: .github/workflows/fast_testing.yml [hunk 8]
```diff
@@ -10,6 +10,35 @@ on:
...
+          echo REPO_LAST_TAG=$(git describe --tags `git rev-list --tags --max-count=1`) >> $GITHUB_ENV
+
+      - name: Hardcoded version relevance check
```
[the remark this answers is not in the document]
Comment (GRISHNOV) [reply .github/workflows/fast_testing.yml:40]: Done

File: queue/init.lua (hunk 5 above)
Comment (GRISHNOV) [reply queue/init.lua]: Done

**Review by DifferentialOrange — COMMENTED** [review]
Sorry, I forgot to push new comments

File: .github/workflows/fast_testing.yml [hunk 9]
```diff
@@ -10,6 +10,40 @@ on:
...
+    steps:
+      - name: Clone the module
+        uses: actions/checkout@v2
+
+      - name: Setup tarantool
+        uses: tarantool/setup-tarantool@v1
+        with:
+          tarantool-version: '2.9'
+
+      - name: Package version
```
Comment (DifferentialOrange) [inline .github/workflows/fast_testing.yml:25]: How should we get through this check on release commit?

File: .github/workflows/fast_testing.yml [hunk 10]
```diff
@@ -10,6 +10,40 @@ on:
...
+        #   https://github.com/actions/checkout/issues/290
+        # But we use 'git describe ..' calls w/o '--tags' flag and it
+        # prevents us from getting the needed tag for packages version
+        # setup. To avoid of it, let's fetch it manually, to be sure
+        # that all tags will exists always. The actions/checkout depth
+        # does not help.
+        run: |
+          git fetch --tags -f
+          tarantoolctl rocks make
+          echo HARDCODED_VERSION=$(tarantool -e "print(require('queue')._VERSION)")
```
Comment (DifferentialOrange) [inline .github/workflows/fast_testing.yml:37]: Does print required here?

Comment (DifferentialOrange) [inline .github/workflows/fast_testing.yml:37]: And I don't see any reason to describe version here as "hardcoded" too. The only thing where this work is relevant is a implementation description in the commit message.

File: .github/workflows/fast_testing.yml [hunk 11]
```diff
@@ -10,6 +10,40 @@ on:
...
+          git fetch --tags -f
+          tarantoolctl rocks make
+          VERSION=$(tarantool -e "print(require('queue')._VERSION)")
+          REPO_LAST_TAG=$(git describe --tags `git rev-list --tags --max-count=1`)
+          if [ "$VERSION" != "$REPO_LAST_TAG" ]; then
+            echo VERSION from code is $VERSION
+            echo REPO_LAST_TAG is $REPO_LAST_TAG
+            echo "Version from code and the last repository tag are not equal"
+            echo "You may have forgotten to update the value in the version.lua file"
+          exit 1
```
Comment (DifferentialOrange) [inline .github/workflows/fast_testing.yml:44]: Indentation level should be the same as with echo

File: .github/workflows/fast_testing.yml [hunk 12]
```diff
@@ -10,6 +10,48 @@ on:
...
+  version_check:
+    runs-on: ubuntu-latest
+
+    strategy:
+      fail-fast: false
+      matrix:
+        tarantool:
+          - '1.10'
+          - '2.8'
+          - '2.10'
```
Comment (DifferentialOrange) [inline .github/workflows/fast_testing.yml:22]: I think you can install only latest here.

File: .github/workflows/fast_testing.yml [hunk 13]
```diff
@@ -10,6 +10,40 @@ on:
...
 jobs:
+  version_check:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Clone the module
+        uses: actions/checkout@v2
+
+      - name: Setup tarantool
+        uses: tarantool/setup-tarantool@v1
```
Comment (DifferentialOrange) [inline .github/workflows/fast_testing.yml:21]: "Node.js 12 actions are deprecated. Please update the following actions to use Node.js 16: actions/checkout@v2, tarantool/setup-tarantool@v1."

Comment (DifferentialOrange) [reply .github/workflows/fast_testing.yml:21]: checkout v3 and setup-tarantool v2 are based on Node 16

File: .github/workflows/fast_testing.yml [hunk 14]
```diff
@@ -10,6 +10,40 @@ on:
...
+        with:
+          tarantool-version: '2.9'
```
[the remark this answers is not in the document]
Comment (GRISHNOV) [reply .github/workflows/fast_testing.yml]: Done with 2.10 (2.9 version is used below in `run_tests` job)

File: .github/workflows/fast_testing.yml (hunk 10 above)
Comment (GRISHNOV) [reply .github/workflows/fast_testing.yml]: Deleted

Comment (GRISHNOV) [reply .github/workflows/fast_testing.yml]: Deleted all `hardcoded` mentions

File: .github/workflows/fast_testing.yml (hunk 11 above)
Comment (GRISHNOV) [reply .github/workflows/fast_testing.yml]: Done

File: .github/workflows/fast_testing.yml (hunk 12 above)
Comment (GRISHNOV) [reply .github/workflows/fast_testing.yml]: Done with 2.10

File: .github/workflows/fast_testing.yml [hunk 15]
```diff
@@ -10,6 +10,40 @@ on:
...
+        # setup. To avoid of it, let's fetch it manually, to be sure
+        # that all tags will exists always. The actions/checkout depth
+        # does not help.
+        run: |
+          git fetch --tags -f
+          tarantoolctl rocks make
+          VERSION=$(tarantool -e "print(require('queue')._VERSION)")
+          REPO_LAST_TAG=$(git describe --tags `git rev-list --tags --max-count=1`)
+          if [ "$VERSION" != "$REPO_LAST_TAG" ]; then
+            echo VERSION from code is $VERSION
```
Comment (DifferentialOrange) [inline .github/workflows/fast_testing.yml:40]: I think you may extract these two lines from `if fi` block since they are useful.

File: .github/workflows/fast_testing.yml (hunk 10 above)
Comment (DifferentialOrange) [reply .github/workflows/fast_testing.yml]: I still see it

File: .github/workflows/fast_testing.yml (hunk 15 above)
Comment (GRISHNOV) [reply .github/workflows/fast_testing.yml]: Done

File: .github/workflows/fast_testing.yml [hunk 16]
```diff
@@ -10,6 +10,40 @@ on:
       - '*'
 
 jobs:
+  version_check:
```
Comment (DifferentialOrange) [inline .github/workflows/fast_testing.yml:13]: Release publish should depend on this one.

Comment (DifferentialOrange) [reply .github/workflows/fast_testing.yml:13]: We have code duplication now.

Comment (DifferentialOrange) [reply .github/workflows/fast_testing.yml:13]: Just remove it

Comment (GRISHNOV) [reply .github/workflows/fast_testing.yml]: Done, now there is only in `publish.yml`

File: .github/workflows/version_checking.yml [hunk 17]
```diff
@@ -0,0 +1,39 @@
+name: version_checking
```
Comment (DifferentialOrange) [inline .github/workflows/version_checking.yml:1]: It shouldn't be a workflow, see tarantool/tarantool actions
https://github.com/tarantool/tarantool/tree/master/.github/actions/environment

File: .github/workflows/packaging.yml [hunk 18]
```diff
@@ -40,6 +40,11 @@ jobs:
       DIST: ${{ matrix.platform.dist }}
 
     steps:
+      - name: Version check
+        # We need this step to run only on push with tag.
+        if: ${{ github.event_name == 'push' && startsWith(github.ref, 'refs/tags/') }}
+        uses: ./.github/actions/version_checking
```
Comment (DifferentialOrange) [inline .github/workflows/packaging.yml:46]: Well, if you use it this way, module clone and tarantool install are redundant.

Comment (GRISHNOV) [reply .github/workflows/packaging.yml:49]: Do you mean delete steps `Clone the module` and `Setup tarantool` from `version_checking/action.yml`?

File: .github/actions/version_checking/action.yml [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+runs:
+  using: "composite"
+  steps:
+    - name: Clone the module
+      uses: actions/checkout@v2
+
+    - name: Setup tarantool
+      uses: tarantool/setup-tarantool@v1
+      with:
+        tarantool-version: '2.9'
```
Comment (DifferentialOrange) [inline .github/actions/version_checking/action.yml:13]: You use old code here without our review fixes

File: .github/workflows/packaging.yml (hunk 18 above)
Comment (DifferentialOrange) [reply .github/workflows/packaging.yml:49]: Feel free to reorder existing CI steps

File: .github/actions/version_checking/action.yml [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+      # right it deannotates the testing tag, check:
+        #  https://github.com/actions/checkout/issues/290
+      # But we use 'git describe ..' calls w/o '--tags' flag and it
+      # prevents us from getting the needed tag for packages version
+      # setup. To avoid of it, let's fetch it manually, to be sure
+      # that all tags will exists always. The actions/checkout depth
+      # does not help.
+      run: |
+        git fetch --tags -f
+          tarantoolctl rocks make
```
Comment (DifferentialOrange) [inline .github/actions/version_checking/action.yml:26]: Indentation is broken here

File: .github/actions/version_checking/action.yml [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+        VERSION=$(tarantool -e "print(require('queue')._VERSION)")
+        REPO_LAST_TAG=$(git describe --tags `git rev-list --tags --max-count=1`)
+        echo VERSION from code is $VERSION
+        echo REPO_LAST_TAG is $REPO_LAST_TAG
+        if [ "$VERSION" != "$REPO_LAST_TAG" ]; then
+          echo "Version from code and the last repository tag are not equal"
+          echo "You may have forgotten to update the value in the version.lua file"
+            exit 1
```
Comment (DifferentialOrange) [inline .github/actions/version_checking/action.yml:34]: And indentation is broken here too

File: .github/workflows/packaging.yml [hunk 22]
```diff
@@ -40,6 +40,14 @@ jobs:
       DIST: ${{ matrix.platform.dist }}
 
     steps:
+      - name: Prepare checkout
```
Comment (DifferentialOrange) [inline .github/workflows/packaging.yml:43]: It's always good to self-review your PR, at least briefly. here we definitely have a traces of the old approach.

File: .github/actions/version_checking/action.yml (hunk 20 above)
Comment (DifferentialOrange) [inline .github/actions/version_checking/action.yml:26]: We need to be careful with this step since it may affect packaging.

Comment (DifferentialOrange) [inline .github/actions/version_checking/action.yml:26]: To be honest, I'm thinking that it would be better as a separate job rather than step. In this case, we'll need to rework queue CD to have separate "build a package" and "publish a package" workflows. It is a bit of extra work. Also it is troublesome for Enterprise CD since we cannot use GitHub artifacts due to limits for private repos. We'll need to think about this question.

File: .github/actions/version_checking/action.yml (hunk 19 above)
Comment (GRISHNOV) [reply .github/actions/version_checking/action.yml]: Now 2.10

File: .github/actions/version_checking/action.yml (hunk 20 above)
Comment (GRISHNOV) [reply .github/actions/version_checking/action.yml]: Done

File: .github/actions/version_checking/action.yml (hunk 21 above)
Comment (GRISHNOV) [reply .github/actions/version_checking/action.yml]: Done

File: .github/workflows/publish.yml [hunk 23]
```diff
@@ -11,6 +11,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
+      - uses: ./.github/actions/version_checking
```
Comment (DifferentialOrange) [inline .github/workflows/publish.yml:14]: You don't need it here

Comment (GRISHNOV) [reply .github/workflows/publish.yml]: Deleted

File: .github/workflows/packaging.yml (hunk 22 above)
Comment (GRISHNOV) [reply .github/workflows/packaging.yml]: Deleted this step, now `Version check` after `Clone the module`

File: .github/actions/version_checking/action.yml [hunk 24]
```diff
@@ -0,0 +1,54 @@
+name: version_checking
```
Comment (DifferentialOrange) [inline .github/actions/version_checking/action.yml:1]: Together with tarantool/actions maintainer we decided to use `check-module-version ` name

File: .github/actions/version_checking/action.yml [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+      with:
+        tarantool-version: '2.10'
+
+    - name: Make rock
+      run: tarantoolctl rocks make
+      shell: bash
+
+    - name: Code version
+      run: |
+        VERSION=$(tarantool -e "print(rawget(require('${{ inputs.module-name }}'), '_VERSION'))")
```
Comment (DifferentialOrange) [inline .github/actions/version_checking/action.yml:23]: Why do you use rawget here? Does `-e` enable strict? For me it works ok.

File: .github/actions/version_checking/action.yml [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+        echo Module name is ${{ inputs.module-name }}
+        echo Detected version from code is $VERSION
+        if [ "$VERSION" = "nil" ]; then
+          echo "Check the version value implementation in the Lua module's code"
+          exit 1
+        fi
+        echo "VERSION=$VERSION" >> $GITHUB_ENV
```
Comment (DifferentialOrange) [inline .github/actions/version_checking/action.yml:30]: Let's use MODULE_VERSION

File: .github/workflows/packaging.yml [hunk 27]
```diff
@@ -3,6 +3,18 @@ name: packaging
 on: [push, pull_request]
 
 jobs:
+  version_check:
+    runs-on: ubuntu-18.04
+    steps:
+      # Required to the version_checking action running.
```
Comment (DifferentialOrange) [inline .github/workflows/packaging.yml:9]: Let's put checkout to version_checking then

File: .github/workflows/packaging.yml [hunk 28]
```diff
@@ -76,19 +89,15 @@ jobs:
             -X PUT ${RWS_URL_PART}/${OS}/${DIST} \
             -u ${RWS_AUTH} \
             -F product=${PRODUCT_NAME}"
-
```
Comment (DifferentialOrange) [inline .github/workflows/packaging.yml:79]: Let's not change unrelated code

File: .github/workflows/packaging.yml [hunk 29]
```diff
@@ -3,6 +3,18 @@ name: packaging
 on: [push, pull_request]
 
 jobs:
+  version_check:
+    runs-on: ubuntu-18.04
```
Comment (DifferentialOrange) [inline .github/workflows/packaging.yml:7]: > The ubuntu-18.04 environment is deprecated, consider switching to ubuntu-20.04(ubuntu-latest), or ubuntu-22.04 instead.

Let's use 20.04 disregarding the version used in other pipelines

File: .github/workflows/publish.yml [hunk 30]
```diff
@@ -9,8 +9,10 @@ jobs:
   publish-rockspec-scm-1:
     if: github.ref == 'refs/heads/master'
     runs-on: ubuntu-latest
+    needs: version_check
```
Comment (DifferentialOrange) [inline .github/workflows/publish.yml:12]: As far as I remember GitHub actions, you can't depend on another workflow job like this. For now it hadn't failed because this workflow doesn't run on PRs.

https://stackoverflow.com/questions/58457140/dependencies-between-workflows-on-github-actions

You need to create another job here. The same as in packaging. A bit of copypaste, yeah, but since it's like three lines of code (after you migrate checkout), it shouldn't be an issue.

File: .github/workflows/publish.yml [hunk 31]
```diff
@@ -9,8 +9,10 @@ jobs:
...
     steps:
       - uses: actions/checkout@v3
+      - uses: ./.github/actions/version_checking
```
Comment (DifferentialOrange) [inline .github/workflows/publish.yml:15]: Now you need to remove this one

File: .github/workflows/publish.yml [hunk 32]
```diff
@@ -19,9 +21,10 @@ jobs:
   publish-rockspec-tag:
     if: startsWith(github.ref, 'refs/tags/')
     runs-on: ubuntu-latest
+    needs: version_check
     steps:
       - uses: actions/checkout@v3
-
+      - uses: ./.github/actions/version_checking
```
Comment (DifferentialOrange) [inline .github/workflows/publish.yml:27]: Now you need to remove this one

**Review by DifferentialOrange — COMMENTED** [review]
Thank you for the updates! After fixing the comments below, our final step in this PR is to migrate the action to tarantool/actions.

File: .github/actions/version_checking/action.yml (hunk 25 above)
Comment (GRISHNOV) [reply .github/actions/version_checking/action.yml:23]: So, I had some problems due to `__index` [here](https://github.com/tarantool/queue/blob/master/queue/init.lua#L42).
I tried to get `_VERSION` value with `pcall`, but it did not work for `__index` when `_VERSION` is `nil`

File: .github/workflows/packaging.yml (hunk 27 above)
Comment (GRISHNOV) [reply .github/workflows/packaging.yml]: Do you mean something like this:

packaging.yml
```Yml
jobs:
  version_check:
    runs-on: ubuntu-18.04
    steps:
      # # Required to the version_checking action running.
      # - name: Clone the module
      #   uses: actions/checkout@v3

      - name: Run action
        uses: ./.github/actions/version_checking
        with:
          module-name: 'queue'
```

Action.yml
```Yml
runs:
  using: "composite"
  steps:
    # Added
    - name: Clone the module
      uses: actions/checkout@v3

    - name: Setup tarantool
      uses: tarantool/setup-tarantool@v2
      with:
        tarantool-version: '2.10'

...
...
...
```

I tried, but [version_check](https://github.com/tarantool/queue/actions/runs/4343172137/jobs/7584884131#step:2:1):
```
Can't find 'action.yml', 'action.yaml' or 'Dockerfile' under '/home/runner/work/queue/queue/.github/actions/version_checking'. 
Did you forget to run actions/checkout before running your local action?
```

File: .github/actions/version_checking/action.yml (hunk 24 above)
Comment (GRISHNOV) [reply .github/actions/version_checking/action.yml]: Done, now it is `.github/actions/check-module-version/action.yml` with `name: Check module version ` inside

File: .github/actions/version_checking/action.yml (hunk 26 above)
Comment (GRISHNOV) [reply .github/actions/version_checking/action.yml]: Done

File: .github/workflows/packaging.yml (hunk 29 above)
Comment (GRISHNOV) [reply .github/workflows/packaging.yml]: Done, now using `20.04` for all `version-check`  jobs

File: .github/workflows/publish.yml (hunk 30 above)
Comment (GRISHNOV) [reply .github/workflows/publish.yml]: Done, implemented job `version-check` in publish.yml

> A bit of copypaste, yeah, but since it's like three lines of code (after you migrate checkout), it shouldn't be an issue.

Related [with](https://github.com/tarantool/queue/pull/199#discussion_r1126262069)

File: .github/workflows/publish.yml (hunk 31 above)
Comment (GRISHNOV) [reply .github/workflows/publish.yml]: Deleted

File: .github/workflows/publish.yml (hunk 32 above)
Comment (GRISHNOV) [reply .github/workflows/publish.yml]: Deleted

File: .github/actions/version_checking/action.yml (hunk 25 above)
Comment (DifferentialOrange) [reply .github/actions/version_checking/action.yml]: Well, `require('queue')._VERSION` should work even before box.cfg

File: .github/workflows/packaging.yml (hunk 28 above)
Comment (GRISHNOV) [reply .github/workflows/packaging.yml]: Fixed

File: .github/actions/version_checking/action.yml (hunk 25 above)
Comment (DifferentialOrange) [reply .github/actions/version_checking/action.yml]: > So, I had some problems due to `__index` [here](https://github.com/tarantool/queue/blob/master/queue/init.lua#L42). I tried to get `_VERSION` value with `pcall`, but it did not work for `__index` when `_VERSION` is `nil`

Well, I think it's okay if `require('queue')._VERSION` would fail with error before our comparison if something is wrong.

File: .github/workflows/packaging.yml (hunk 27 above)
Comment (DifferentialOrange) [reply .github/workflows/packaging.yml]: Oh, I get it now. It should be okay after you extract action to a separate action repo.

File: .github/actions/version_checking/action.yml (hunk 25 above)
Comment (GRISHNOV) [reply .github/actions/version_checking/action.yml]: Done

File: .github/workflows/packaging.yml (hunk 27 above)
Comment (GRISHNOV) [reply .github/workflows/packaging.yml]: Now using:

```yml
jobs:
  version-check:
    runs-on: ubuntu-20.04
    steps:
      - name: Run action
        uses: tarantool/actions/check-module-version@igrishnov/gh-204-add-hardcoded-versioning-support
        with:
          module-name: 'queue'
          repository: 'tarantool/queue'
```

The `@igrishnov/gh-204-add-hardcoded-versioning-support` will be deleted

Also I added `needs: version-check` to `package` job of `packaging.yml` and deleted local `action.yml`

File: .github/workflows/packaging.yml [hunk 33]
```diff
@@ -11,7 +20,8 @@ jobs:
       github.event.pull_request.head.repo.full_name != github.repository
     # Packaging for CentOS 7 does not work with other versions, see:
     # https://github.com/packpack/packpack/issues/145
-    runs-on: ubuntu-18.04
+    runs-on: ubuntu-20.04
```
Comment (DifferentialOrange) [inline .github/workflows/packaging.yml:23]: Please, move all maintenance version updates to a separate commit.

File: .github/workflows/publish.yml [hunk 34]
```diff
@@ -19,9 +29,9 @@ jobs:
   publish-rockspec-tag:
     if: startsWith(github.ref, 'refs/tags/')
     runs-on: ubuntu-latest
+    needs: version-check
     steps:
       - uses: actions/checkout@v3
-
```
Comment (DifferentialOrange) [inline .github/workflows/publish.yml:24]: This update also seems redundant

File: .github/workflows/packaging.yml [hunk 35]
```diff
@@ -3,6 +3,15 @@ name: packaging
...
 
 jobs:
+  version-check:
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Run action
+        uses: tarantool/actions/check-module-version@igrishnov/gh-204-add-hardcoded-versioning-support
+        with:
+          module-name: 'queue'
+          repository: 'tarantool/queue'
```
Comment (DifferentialOrange) [inline .github/workflows/packaging.yml:13]: How checkout works? We need to specify the repo inside the action, but doesn't need to specify PR commit?

**Review by DifferentialOrange — APPROVED** [review]
Thank you for the updates, this version seems like a good and rather small solution.

File: .github/workflows/packaging.yml (hunk 35 above)
Comment (GRISHNOV) [reply .github/workflows/packaging.yml:13]: AFAIU, commit can be [specified](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#runsstepsuses), but we can just use `@master` instead `@igrishnov/gh-204-add-hardcoded-versioning-support` here after done [PR](https://github.com/tarantool/actions/pull/54)

File: .github/workflows/publish.yml (hunk 34 above)
Comment (GRISHNOV) [reply .github/workflows/publish.yml:24]: Fixed

File: .github/workflows/packaging.yml (hunk 33 above)
Comment (GRISHNOV) [reply .github/workflows/packaging.yml:23]: Done, rebased to `master` that has this change

File: .github/workflows/packaging.yml (hunk 35 above)
Comment (DifferentialOrange) [reply .github/workflows/packaging.yml:13]: No, I mean the checkout inside the action. Why we need to specify `tarantool/queue` repo, but do not required to specify commit to checkout?

Comment (GRISHNOV) [reply .github/workflows/packaging.yml]: It looks like the `repository` variable can be [removed](https://github.com/actions/checkout/blob/main/README.md?plain=1#L24-L31), deleted

File: .github/workflows/packaging.yml [hunk 36]
```diff
@@ -3,6 +3,14 @@ name: packaging
 on: [push, pull_request]
 
 jobs:
+  version-check:
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Run action
```
Comment (DifferentialOrange) [inline .github/workflows/packaging.yml:9]: Let's add the same

```
    # Skip pull request jobs when the source branch is in the same
    # repository.
    if: |
      github.event_name == 'push' ||
      github.event.pull_request.head.repo.full_name != github.repository
```

skip so there won't a double job execution for each PR push

Comment (GRISHNOV) [reply .github/workflows/packaging.yml:14]: Done, added

File: .github/workflows/packaging.yml [hunk 37]
```diff
@@ -3,6 +3,19 @@ name: packaging
...
 jobs:
+  version-check:
+    # Skip pull request jobs when the source branch is in the same
+    # repository.
+    if: |
+      github.event_name == 'push' ||
+      github.event.pull_request.head.repo.full_name != github.repository
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Run action
```
Comment (DifferentialOrange) [inline .github/workflows/packaging.yml:14]: "Check module version" sounds more useful

Comment (GRISHNOV) [reply .github/workflows/packaging.yml]: Done, now `name:` is `Check module version`

**Review by LeonidVas — APPROVED** [review]
Hi! Thank you for the patch.
LGTM.

File: .github/workflows/packaging.yml [hunk 38]
```diff
@@ -3,6 +3,19 @@ name: packaging
 on: [push, pull_request]
 
 jobs:
+  version-check:
```
[the remark this answers is not in the document]
Comment (DifferentialOrange) [reply .github/workflows/packaging.yml:6]: Please, add a comment "Run not only on tags, otherwise dependent job will skip" or something like this

[the remark this answers is not in the document]
Comment (GRISHNOV) [reply .github/workflows/packaging.yml:7]: Done, added

[the remark this answers is not in the document]
Comment (DifferentialOrange) [reply .github/workflows/packaging.yml:6]: I don't see it

File: .github/workflows/publish.yml [hunk 39]
```diff
@@ -6,9 +6,20 @@ on:
...
+        # We need this step to run only on push with tag.
+        if: ${{ github.event_name == 'push' && startsWith(github.ref, 'refs/tags/') }}
+        uses: tarantool/actions/check-module-version@igrishnov/gh-204-add-hardcoded-versioning-support
+        with:
+          module-name: 'queue'
+
   publish-rockspec-scm-1:
     if: github.ref == 'refs/heads/master'
     runs-on: ubuntu-latest
+    needs: version-check
```
Comment (DifferentialOrange) [inline .github/workflows/publish.yml:22]: Let's remove it here since it doesn't do anything

File: .github/workflows/publish.yml [hunk 40]
```diff
@@ -6,9 +6,20 @@ on:
     tags: ['*']
 
 jobs:
+  version-check:
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Check module version
+        # We need this step to run only on push with tag.
+        if: ${{ github.event_name == 'push' && startsWith(github.ref, 'refs/tags/') }}
```
Comment (DifferentialOrange) [inline .github/workflows/publish.yml:14]: And then we can use `if` condition on job-level here

File: .github/workflows/packaging.yml (hunk 38 above)
[the remark this answers is not in the document]
Comment (GRISHNOV) [reply .github/workflows/packaging.yml:7]: Updated

File: .github/workflows/publish.yml (hunk 39 above)
Comment (GRISHNOV) [reply .github/workflows/publish.yml]: Deleted

File: .github/workflows/publish.yml (hunk 40 above)
Comment (GRISHNOV) [reply .github/workflows/publish.yml]: Moved to job-level

File: .github/workflows/packaging.yml [hunk 41]
```diff
@@ -3,6 +3,22 @@ name: packaging
...
+    # repository.
+    if: |
+      github.event_name == 'push' ||
+      github.event.pull_request.head.repo.full_name != github.repository
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Check module version
+        # We need this step to run only on push with tag.
+        if: ${{ github.event_name == 'push' && startsWith(github.ref, 'refs/tags/') }}
+        uses: tarantool/actions/check-module-version@igrishnov/gh-204-add-hardcoded-versioning-support
```
Comment (DifferentialOrange) [inline .github/workflows/packaging.yml:18]: Let's change these one and merge it

Comment (GRISHNOV) [reply .github/workflows/packaging.yml]: Done
